spring-cloud-gcp-test/src/main/java/org/springframework/cloud/gcp/test/PubSubEmulatorHelper.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2017-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.test;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class PubSubEmulatorHelper extends AbstractEmulatorHelper {
```

Comment:
Can we please refactor this not to use inheritance?
You can try turning all of the abstract methods of `AbstractEmulatorHelper` into an interface that each emulator would implement and then a final `EmulatorHelper` can be customized by.
Alternatively, you can make `EmulatorHelper` a generic helper that any implementation of an emulator can re-use.

Reply:
Why though? Sincerely curious; this seems like a reasonable case for inheritance.

Reply:
I'm sure if you search for "inheritance is evil", lots of reasons will pop up.
From Effective Java 3rd edition:
See:
Item 18: Favor composition over inheritance
Item 19: Design and document for inheritance or else prohibit it 
Item 20: Prefer interfaces to abstract classes

Reply:
I don't do religious wars.

Reply:
Do you think it's the same as tabs vs spaces? 😄

Reply:
Yep. Each side can be justified with rational arguments, and yet, the disagreement lives on. We'll have to do a PR converting Spring tabs into Google spaces soon...